tfhe/examples/regex_engine/engine.rs

```diff
@@ -0,0 +1,268 @@
+                    .flat_map(|(branch_prev, branch_prev_c_pos)| {
+                        build_branches(content, re_x, branch_prev_c_pos)
+                            .into_iter()
+                            .map(move |(branch_x, branch_x_c_pos)| {
+                                let branch_prev = branch_prev.clone();
+                                (
+                                    Rc::new(move |exec: &mut Execution| {
+                                        let res_prev = branch_prev(exec);
+                                        let res_x = branch_x(exec);
+                                        exec.ct_and(res_prev, res_x)
+                                    }) as LazyExecution,
+                                    branch_x_c_pos,
+                                )
+                            })
+                    })
+                    .collect()
+            },
+        ),
+        _ => panic!("unmatched regex variant"),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::engine::has_match;
+    use test_case::test_case;
+
+    use tfhe::integer::{ServerKey, RadixClientKey};
+    use crate::ciphertext::{encrypt_str, gen_keys, StringCiphertext};
+    use bincode;
+    use lazy_static::lazy_static;
+    use std::io::Write;
+
+    lazy_static! {
+        pub static ref KEYS: (RadixClientKey, ServerKey) = gen_keys();
+    }
+
+    #[test_case("ab", "/ab/", 1)]
+    #[test_case("b", "/ab/", 0)]
+    #[test_case("ab", "/a?b/", 1)]
+    #[test_case("b", "/a?b/", 1)]
+    #[test_case("ab", "/^ab|cd$/", 1)]
+    #[test_case(" ab", "/^ab|cd$/", 0)]
+    #[test_case(" cd", "/^ab|cd$/", 0)]
+    #[test_case("cd", "/^ab|cd$/", 1)]
+    #[test_case("abcd", "/^ab|cd$/", 0)]
+    #[test_case("abcd", "/ab|cd$/", 1)]
+    #[test_case("abc", "/abc/", 1)]
+    #[test_case("123abc", "/abc/", 1)]
+    #[test_case("123abc456", "/abc/", 1)]
+    #[test_case("123abdc456", "/abc/", 0)]
+    #[test_case("abc456", "/abc/", 1)]
+    #[test_case("bc", "/a*bc/", 1)]
+    #[test_case("cdaabc", "/a*bc/", 1)]
+    #[test_case("cdbc", "/a+bc/", 0)]
+    #[test_case("bc", "/a+bc/", 0)]
+    #[test_case("Ab", "/ab/i", 1 ; "ab case insensitive")]
+    #[test_case("Ab", "/ab/", 0 ; "ab case sensitive")]
+    #[test_case("cD", "/ab|cd/i", 1)]
+    #[test_case("cD", "/cD/", 1)]
+    #[test_case("de", "/^ab|cd|de$/", 1 ; "multiple or")]
+    #[test_case(" de", "/^ab|cd|de$/", 0 ; "multiple or nests below ^")]
+    fn test_has_match(content: &str, pattern: &str, exp: u64) {
+        // it would be more correct to encrypt the content properly with the
+        // client key, however in order to significantly speed up these test
+        // cases we're encoding them trivially into ciphertext
+        let ct_content: StringCiphertext = content
+            .as_bytes()
+            .iter()
+            .map(|byte| KEYS.1.create_trivial_radix(*byte as u64, 4))
+            .collect();
+        let ct_res = has_match(&KEYS.1, &ct_content, pattern).unwrap();
```

Comment:
Here, we want a real encryption using the client key, so that if we change parameters in PARAM_MESSAGE_2_CARRY_2 would would know it that breaks the regex example.

And using real encryption does not seem to be that slower, well its 2x slower, but it only goes from 2sec to 4sec so its fine